content/en/blog/_posts/2023-12-13-Canary-Deployment-With-Kubernetes/index.md

```diff
@@ -0,0 +1,282 @@
+date: 2023-12-13
+slug: canary-deployment-with-kubernetes
+evergreen: true
+---
+
+**Author**: Mohammed Affan
+
+## Introduction:
+
+Welcome to a comprehensive guide to canary deployment with kubernetes.
+In modern software development getting the quick feedback form the users
+is crucial for delivering a high quality product and implementing agile
+delivery. Developers try to get feedback from the users without ruining the
+user experience. That's where canary deployment comes into the picture.
+
+## Prerequisites:
+
+- You need to have a Kubernetes cluster (minikube would also work).
+- You also need to have `kubectl` installed on the local machine - see [install tools](/docs/tasks/tools/#kubectl).
+
+## What is a canary deployment?
+
+Suppose you have a web application and you want to implement a new feature.
+But you are not sure if your users are going to like this feature or not.
+This makes testing and getting user feedback risky. A solution to mitigate
+this risk is to test the new feature with a limited number of users.
+
+Here is an example showing canary deployment. Assume that you have two versions of an app. Named App v1
+and App v2 respectively. Currently only the App v1 is deployed to the production. To fully deploy the
+App v2 to the production first you need to gather some user feedback. To gather the user
+feedback without ruining the user experience you decided to send the 90% of your traffic
+to the App v1 and 10% of the traffic to the App v2. This type of weighted deployment
+method is known as the canary deployment.
+
+{{< figure src="canary-deployment.png" alt="Illustration of canary deployment where 90% of the traffic is going towards the app v1 and the 10% of the traffic going towards the app v2" >}}
+
+## Deploying the app:
+
+Let's deploy the stable and canary versions of the app. The
+[canary-app](https://hub.docker.com/repository/docker/affan7/canary-app/general)
+repository contains two docker images. One is the canary-app:v1 which is the
+stable version of the app. And the other image is canary-app:v2 which is the
+canary version of the app.
+
+To deploy the app, create a `deployment.yaml` manifest file with the following
+configuration.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: canary-app-deployment
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: canary-app
+      version: v1
+  template:
+    metadata:
+      labels:
+        app: canary-app
+        version: v1
+    spec:
+      containers:
+      - name: canary-app
+        image: affan7/canary-app:v1
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: canary-app-deployment-v2
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: canary-app
+      version: v2
+  template:
+    metadata:
+      labels:
+        app: canary-app
+        version: v2
+    spec:
+      containers:
+      - name: canary-app
+        image: affan7/canary-app:v2
```

Comment:
- Why this image?
- How can we be sure this image will be maintained indefinitely?